src/Server.cpp

```diff
@@ -673,12 +710,88 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
+ * and the connection's HTTP status is set to StatusRequestEntityTooLarge.
+ * If the complete request body is received, it is parsed and the connection's status is set to BuildResponse.
+ * If an exception occurs during parsing, an error is logged and the connection's status is set to BuildResponse.
+ *
+ * @param server The server object.
+ * @param clientFd The file descriptor of the client socket.
+ * @param connection The connection object.
+ */
 void connectionReceiveBody(Server& server, int clientFd, Connection& connection)
 {
 	LOG_DEBUG << "ReceiveBody for: " << connection.m_clientSocket;
 
-	(void)server;
-	(void)clientFd;
+	char buffer[Server::s_clientBodyBufferSize] = {};
+
+	const ssize_t bytesRead = server.readFromSocket(clientFd, buffer, sizeof(buffer), 0);
+	if (bytesRead == -1) {
+		LOG_DEBUG << "Internal server error while reading from socket: " << connection.m_clientSocket;
+		close(clientFd);
+		connection.m_status = Connection::Closed;
+	} else if (bytesRead == 0) {
+		LOG_DEBUG << "Connection closed by client: " << connection.m_clientSocket;
+		close(clientFd);
+		connection.m_status = Connection::Closed;
+	} else {
+		connection.m_buffer += buffer;
+		if (connection.m_buffer.size() >= Server::s_clientMaxBodySize) {
+			LOG_ERROR << "Maximum allowed client request body size reached from " << connection.m_clientSocket;
+			connection.m_request.httpStatus = StatusRequestEntityTooLarge;
+			connection.m_status = Connection::BuildResponse;
+		}
+		try {
+			if (isCompleteBody(connection.m_buffer, connection.m_request)) {
+				LOG_DEBUG << "Received complete request body: " << '\n' << connection.m_buffer;
+				server.parseBody(connection.m_buffer, connection.m_request);
+				connection.m_status = Connection::BuildResponse;
+				server.modifyEvent(clientFd, EPOLLOUT);
+				connection.m_buffer.erase(0);
+			} else
+				LOG_DEBUG << "Received partial request body: " << '\n' << connection.m_buffer;
+		} catch (std::exception& e) {
+			LOG_ERROR << "Error: " << e.what();
+			connection.m_status = Connection::BuildResponse;
+			return;
+		}
+	}
+	connection.m_timeSinceLastEvent = std::time(0);
+}
+
+/**
+ * @brief Checks if the connection buffer contains a complete request body.
+ *
+ * A complete request body is determined by the presence of the "Content-Length" header or the "Transfer-Encoding"
+ * header. If the "Content-Length" header is present, the function checks if the size of the connection buffer matches
+ * the specified content length. If the "Transfer-Encoding" header is present, the function checks if the connection
+ * buffer contains the end marker "0\r\n\r\n".
+ *
+ * @param connectionBuffer The buffer to check for a complete request body.
+ * @param request The HTTPRequest object containing the request headers.
+ * @return true if the buffer contains a complete request body, false otherwise.
+ */
+bool isCompleteBody(const std::string& connectionBuffer, HTTPRequest& request)
```

Comment:
These are two different things:

In the RequestParser we establish whether or not the request has a body and whether or not it is chunked. We save it as a state description in the HTTPRequest.

Then we need to determine whether the received body is complete or not. And here we need to take into consideration different requirements for each case: chunked or non-chunked. That is performed with this function. What you describe is the next step once we have ensured that the body is complete. We call the appropriate method to parse the body. This is done in parseBody():

```
void RequestParser::parseBody(const std::string& bodyString, HTTPRequest& request)
{
	m_requestStream.str(bodyString);
	if (request.isChunked)
		parseChunkedBody(request);
	else
		parseNonChunkedBody(request);
	resetRequestStream();
}
``

Reply:
Why do we need to check for the headers Content-Length and Transfer-Encoding
```
std::map<std::string, std::string>::const_iterator contentLengthIterator
		= connection.m_request.headers.find("Content-Length");
	std::map<std::string, std::string>::const_iterator transferEncodingIterator
		= connection.m_request.headers.find("Transfer-Encoding");
```
If I understood correctly the parser makes sure that the request headers are correct. It also sets 
- hasBody (meaning we have either Content-Length or Transder-Encoding) and 
- isChunked (meaning we have chunked encoding).

So the condition `If (HTTPRequest.hasBody && !HTTPRequest.isChunked)` checks for the case that Content-Length is given and transfer is not chunked (as already determined by the parser). We only need to extract out content length 
`unsigned long contentLength = std::strtoul(connection.m_request.headers.at("Content-Length"), NULL, decimalBase);`

If this condition doesn't fire we know that we have chunked encoding (as the parser has checked). So we can just return `connection.m_buffer.find("0\r\n\r\n") != std::string::npos;`

So we have 
```
If (HTTPRequest.hasBody && !HTTPRequest.isChunked) {
  unsigned long contentLength = std::strtoul(connection.m_request.headers.at("Content-Length"), NULL, decimalBase);
  if (contentLength == connection.m_buffer.size())
    return true;
  if (contentLength < connection.m_buffer.size())
    connection.m_request.httpStatus = StatusBadRequest;
  return false;
} else {
 return connection.m_buffer.find("0\r\n\r\n") != std::string::npos;
} 
```

We could also save the Content-Length in the HTTP Request in the parser, so that we don't have to convert everytime.

Reply:
I see what you mean. Thought you wanted to jump this step completely and go to the parsing directly.

Yeah, we can just use the booleans instead of searching for the map entries. We can even omit the first "HTTPRequest.hasBody" check as we already know in this function call that there is a body. Reduced the function to simple var checks.